scriptworker/cot/verify.py

```diff
@@ -435,36 +435,54 @@ def verify_docker_image_sha(chain, link):
     raise_on_errors(errors)
 
 
-# find_task_dependencies {{{1
-def find_task_dependencies(task, name, task_id):
+def find_sorted_task_dependencies(task, task_name, task_id):
     """Find the taskIds of the chain of trust dependencies of a given task.
 
     Args:
         task (dict): the task definition to inspect.
-        name (str): the name of the task, for logging and naming children.
+        task_name (str): the name of the task, for logging and naming children.
         task_id (str): the taskId of the task.
 
     Returns:
-        dict: mapping dependent task ``name`` to dependent task ``taskId``.
+        list: tuples associating dependent task ``name`` to dependent task ``taskId``.
     """
-    log.info("find_task_dependencies {} {}".format(name, task_id))
+    log.info("find_sorted_task_dependencies {} {}".format(task_name, task_id))
+
+    cot_input_dependencies = [
+        _craft_dependency_tuple(task_name, task_type, task_id)
+        for task_type, task_id in task['extra'].get('chainOfTrust', {}).get('inputs', {}).items()
+    ]
+
+    upstream_artifacts_dependencies = [
+        _craft_dependency_tuple(task_name, artifact_dict['taskType'], artifact_dict['taskId'])
+        for artifact_dict in task.get('payload', {}).get('upstreamArtifacts', [])
+    ]
+
+    dependencies = [*cot_input_dependencies, *upstream_artifacts_dependencies]
+    dependencies = _filter_out_duplicates(dependencies)
+    dependencies = _sort_dependencies_by_name_then_task_id(dependencies)
+
     decision_task_id = get_decision_task_id(task)
-    decision_key = '{}:decision'.format(name)
-    task_type = guess_task_type(name)
-    dep_dict = {}
-    for key, val in task['extra'].get('chainOfTrust', {}).get('inputs', {}).items():
-        dep_dict['{}:{}'.format(name, key)] = val
-    if 'upstreamArtifacts' in task['payload']:
-        upstream_ids = {}
-        for artifact_dict in task['payload']['upstreamArtifacts']:
-            if artifact_dict['taskId'] not in upstream_ids:
-                upstream_ids[artifact_dict['taskId']] = artifact_dict['taskType']
-        for upstream_task_id, upstream_task_type in upstream_ids.items():
-            dep_dict['{}:{}'.format(name, upstream_task_type)] = upstream_task_id
+    task_type = guess_task_type(task_name)
     if decision_task_id != task_id and task_type != 'decision':
-        dep_dict[decision_key] = decision_task_id
-    log.info(dep_dict)
-    return dep_dict
+        # make sure we deal with the decision task first, or we may populate
+        # signing:build0:decision before signing:decision
+        dependencies.insert(0, _craft_dependency_tuple(task_name, 'decision', decision_task_id))
+
+    log.info('found dependencies: {}', dependencies)
```

Comment:
I ran `<venv_path>/bin/verify_cot --task-type signing -- ExoqkVcxQ5ufThL4htdR0A` just to verify this stuff works properly.

This line is throwing errors, e.g.

```
INFO:scriptworker.cot.verify:find_sorted_task_dependencies signing:build:docker-image:decision bswacyRlQlS0JL-IqccLQw
--- Logging error ---
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/logging/__init__.py", line 980, in emit
    msg = self.format(record)
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/logging/__init__.py", line 830, in format
    return fmt.format(record)
  File "/Users/asasaki/src/tc/johan/scriptworker/cot/verify.py", line 1189, in format
    return super(AuditLogFormatter, self).format(record)
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/logging/__init__.py", line 567, in format
    record.message = record.getMessage()
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/logging/__init__.py", line 330, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "jv/bin/verify_cot", line 11, in <module>
    load_entry_point('scriptworker', 'console_scripts', 'verify_cot')()
  File "/Users/asasaki/src/tc/johan/scriptworker/cot/verify.py", line 1281, in verify_cot_cmdln
    loop.run_until_complete(verify_chain_of_trust(cot))
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/base_events.py", line 375, in run_until_complete
    self.run_forever()
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/base_events.py", line 345, in run_forever
    self._run_once()
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/base_events.py", line 1312, in _run_once
    handle._run()
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/events.py", line 125, in _run
    self._callback(*self._args)
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/tasks.py", line 307, in _wakeup
    self._step()
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/Users/asasaki/src/tc/johan/scriptworker/cot/verify.py", line 1212, in verify_chain_of_trust
    await build_task_dependencies(chain, chain.task, chain.name, chain.task_id)
  File "/Users/asasaki/src/tc/johan/scriptworker/cot/verify.py", line 518, in build_task_dependencies
    await build_task_dependencies(chain, task_defn, task_name, task_id)
  File "/Users/asasaki/src/tc/johan/scriptworker/cot/verify.py", line 518, in build_task_dependencies
    await build_task_dependencies(chain, task_defn, task_name, task_id)
  File "/Users/asasaki/src/tc/johan/scriptworker/cot/verify.py", line 518, in build_task_dependencies
    await build_task_dependencies(chain, task_defn, task_name, task_id)
  File "/Users/asasaki/src/tc/johan/scriptworker/cot/verify.py", line 504, in build_task_dependencies
    sorted_dependencies = find_sorted_task_dependencies(task, name, my_task_id)
  File "/Users/asasaki/src/tc/johan/scriptworker/cot/verify.py", line 472, in find_sorted_task_dependencies
    log.info('found dependencies: {}', dependencies)
Message: 'found dependencies: {}'
Arguments: ([],)
--- Logging error ---
```


Instead,

`log.info('found dependencies: {}'.format(dependencies))` works.  I imagine `str(dependencies)` may as well.